### ISSUE TITLE
Snapshot Update

### ISSUE THREAD

**Issue by mlavio829** [potmat/homebridge-google-nest-sdm#45]
Hi,
Wondering if there is any way to add a snapshot refresh option? 
I understand that when motion is triggered is shows me the snapshot but I’m wondering if there is anyway to pull a fresh snapshot every time I look at HomeKit? I have to click the camera video and then it updates. 

Is this possible?

**Comment by potmat** [comment]
Hi @mlavio829,

There's no way for the plugin to distinguish between a snapshot request because you opened the app, and one that happens because the app is already open.  Nor is there any way to add a "button" to refresh the snapshots that I'm aware of.

As mentioned in the FAQ, the only way to get an image on demand is to actually start the video stream and grab the first frame.  This works OK if you have only one camera, but if you have multiple cameras starting all their streams every 10 seconds uses a huge amount of resources and also leads to the SDM API throwing out rate limit errors (you're making too many stream requests).

**Comment by mlavio829** [comment]
Yes I don’t like the generic icon either. The Starlink Hub is able to get the snapshots on the newer Nest doorbells somehow. Not sure how tho.

**Comment by potmat** [comment]
@mlavio829: The Starling Home Hub uses the unofficial APIs as well as the official ones, I would guess that's how they do it.

**Comment by potmat** [comment]
@lnjustin:  Yes that's what we're talking about. I agree the recent change to the home app where you just see a tile of images means you cannot tell which camera is which.  If you click on "Cameras >" then you see each camera with the name overlayed.  I suppose I could add some text into the image with the camera name to aid with the problem.

**Comment by tealtwo** [comment]
> @lnjustin: Yes that's what we're talking about. I agree the recent change to the home app where you just see a tile of images means you cannot tell which camera is which. If you click on "Cameras >" then you see each camera with the name overlayed. I suppose I could add some text into the image with the camera name to aid with the problem.

You could add a button in the config on weather to enable snapshots or not, and if it does get rate limited (which it will thanks google) then too keep the existing snapshot until a new one is pulled. For example, I don't need a snapshot every 2 seconds of the inside of my garage, but I do need the snapshot of my living room or the kids room.

**Comment by noahbm** [comment]
+1 on something like this! 

Allowing the last snapshot image to persist until there is a fresh (event generated) one to replace it would make it easier to identify cameras from the Home Screen / make things a bit more user friendly.

Incredible work btw @potmat! Just finished setting things up. Getting the API credentials sorted was a bit of a journey, but everything on your end worked flawlessly. Cheers!

**Comment by tealtwo** [comment]
@potmat I also noticed on my starling home hub the same thing, the image would either show the last snapshot in HomeKit as 1s old, or it would show a snapshot from the previous 10 seconds so it takes two snapshots for let’s say a delivery man to be seen

**Comment by regg00** [comment]
+1 
that feature would be very appreciated

**Comment by aterriblegamer** [comment]
Got this set up today. Feel like Google just stole $5 from me.the static icons are useless. I used to be able to see all my cameras with the old plugin. I am giving up using Homebridge for Nest. Will be dumping the cameras as well

**Comment by potmat** [comment]
Ideally the API should support a snapshot feature.  Feature requests for the SDM API can be made here: https://issuetracker.google.com/issues?q=status:open%20componentid:1351014&s=created_time:desc

**Comment by jradwan** [comment]
> Ideally the API should support a snapshot feature. Feature requests for the SDM API can be made here: https://issuetracker.google.com/issues?q=status:open%20componentid:1351014&s=created_time:desc

Has anyone already submitted a feature request?

**Comment by ajplotkin** [comment]
The short version: there's no way to get an on-demand fresh snapshot from the SDM API itself, and that's the core limitation — not the plugin.

Nest/SDM gives you only two ways to get a frame:

1. **`CameraEventImage.GenerateImage`** — a real JPEG, but *only for a motion/doorbell event* (it needs an `eventId`). This plugin already uses it: on an event it fetches the image and caches it as `this.image`, which is why the tile briefly shows a real picture right after motion. But it's cleared after **10 seconds** (`setTimeout(() => this.image = null, 10000)`), so it reverts to the logo. There's no event id for "I just opened the Home app," so idle refreshes have nothing to generate from.
2. **A live stream** — you can pull a frame from it, but spinning one up per snapshot is slow and burns the SDM per-device stream quota (which is why the plugin doesn't).

So two realistic paths if you want better tiles:

- **Light, plugin-only:** keep that last event image around much longer — persist it as the tile until the next event/stream instead of nulling it at 10s. Free, no extra API calls; you just get a real (if not-live) tile for a while after any event instead of a logo. It's also worth loosening the 30s event-age gate (`MAX_EVENT_AGE_SECONDS`): SDM event delivery is often 40–60s late, so many events get dropped as "too old" before the image is even fetched.
- **Heavy, real on-demand snapshots:** keep one stream warm per camera and grab frames from it. That needs a warm-stream layer — [go2rtc](https://github.com/AlexxIT/go2rtc) is built for exactly this. A complete guide + installer that wires warm go2rtc snapshots (and optional go2rtc-routed live view) into this plugin is at github.com/ajplotkin/nest-homekit-snapshots, if it's useful as a reference or an upstream direction.

To answer the original question directly: no, there's no SDM call for "fresh snapshot every time you look" — the image you see after tapping into the camera is the live stream starting. The nearest no-go2rtc improvement is persisting the event image longer.

**Comment by tealtwo** [comment]
> The short version: there's no way to get an on-demand fresh snapshot from the SDM API itself, and that's the core limitation — not the plugin.
> 
> Nest/SDM gives you only two ways to get a frame:
> 
> 1. **`CameraEventImage.GenerateImage`** — a real JPEG, but *only for a motion/doorbell event* (it needs an `eventId`). This plugin already uses it: on an event it fetches the image and caches it as `this.image`, which is why the tile briefly shows a real picture right after motion. But it's cleared after **10 seconds** (`setTimeout(() => this.image = null, 10000)`), so it reverts to the logo. There's no event id for "I just opened the Home app," so idle refreshes have nothing to generate from.
> 2. **A live stream** — you can pull a frame from it, but spinning one up per snapshot is slow and burns the SDM per-device stream quota (which is why the plugin doesn't).
> 
> So two realistic paths if you want better tiles:
> 
> - **Light, plugin-only:** keep that last event image around much longer — persist it as the tile until the next event/stream instead of nulling it at 10s. Free, no extra API calls; you just get a real (if not-live) tile for a while after any event instead of a logo. It's also worth loosening the 30s event-age gate (`MAX_EVENT_AGE_SECONDS`): SDM event delivery is often 40–60s late, so many events get dropped as "too old" before the image is even fetched.
> - **Heavy, real on-demand snapshots:** keep one stream warm per camera and grab frames from it. That needs a warm-stream layer — [go2rtc](https://github.com/AlexxIT/go2rtc) is built for exactly this. A complete guide + installer that wires warm go2rtc snapshots (and optional go2rtc-routed live view) into this plugin is at github.com/ajplotkin/nest-homekit-snapshots, if it's useful as a reference or an upstream direction.
> 
> To answer the original question directly: no, there's no SDM call for "fresh snapshot every time you look" — the image you see after tapping into the camera is the live stream starting. The nearest no-go2rtc improvement is persisting the event image longer.
> 


I have already reverse engineered the Nest cams and the WebRTC path and implemented a local streaming path with a bypass so it doesn't go through googles cloud (it went through cloud even with STUN sometimes oddly when recording to NVR) as well as moved past Nest cams to a local UniFi setup and recommend anyone else do the same. Google Nest cameras are utter garbage hardware and frankly if you want HomeKit support go buy a competent camera Google isn't worth the workarounds (that have to get patched when they randomly break the api, and rate limit you so heavily without a f**cking enterprise ID which they won't issue to literally anyone for some reason) when you can get a cheaper better product elsewhere, especially now that Google home is starting to allow third party cameras. 


The plugins current capabilities are fine given it gets rate limited on 3 days of using it anyway.
